Autoconf xlib checks

psychon opened the pull request:
From the commit message:

    Build tests conditionally
    
    This moves the check for x11 and x11-xcb to configure.ac. The tests are only
    enabled if the necessary libraries are found. Additionally, this commit adds
    --enable-tests / --disable-tests flags to configure to forcefully disable tests
    and, more important, to make configure fail if the dependencies for the tests
    could not be found.
    
    I hope someone can tell me a nicer way to implement this...

(The first commit is just there since I noticed only one test actually needs xcb-aux. I still left the dependency in `Makefile.am`.)

@stapelberg Can you somehow easily test this? I guess you'd want `--enable-tests` to have configure fail if some test-dependencies could not be found.

Airblader commented:
Thank you for this. :) Since I have little to no knowledge on autoconf, I'll wait to see if @stapelberg can maybe say something about it.

stapelberg commented:
Tested it:

* Working as expected: Not having libx11-dev installed and running configure with default arguments results in no tests.
* Working as expected: Not having libx11-dev installed and running configure with --enable-tests results in a failed pkg-config check.
* Not working as expected: Having libx11-dev installed and running configure with --enable-tests results in a build error:

```
/bin/bash ./libtool  --tag=CC   --mode=link gcc -Wall -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Wlogical-op -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -g -O2 -fstack-protector-strong -Wformat -Werror=format-security  -Wl,-z,relro -o tests/tests_database tests/tests_tests_database-tests_utils.o tests/tests_tests_database-tests_database.o libxcb-xrm.la -lxcb -lX11-xcb -lX11 -lxcb 
libtool: link: gcc -Wall -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Wlogical-op -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wl,-z -Wl,relro -o tests/.libs/tests_database tests/tests_tests_database-tests_utils.o tests/tests_tests_database-tests_database.o  ./.libs/libxcb-xrm.so -lX11-xcb -lX11 -lxcb
/usr/bin/ld: tests/tests_tests_database-tests_database.o: undefined reference to symbol 'xcb_aux_get_screen'
//usr/lib/x86_64-linux-gnu/libxcb-util.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

psychon commented:
Whoops, I lost xcb-aux from the list of libraries to link against. I fixed that (and cleaned up the overlinking a bit... this is starting to look really messy).

stapelberg commented:
Tested it again:

* Working as expected: Not having libx11-dev installed and running configure with default arguments results in no tests.
* Working as expected: Not having libx11-dev installed and running configure with --enable-tests results in a failed pkg-config check.
* Working as expected: Having libx11-dev installed and running configure with --enable-tests works.

Thanks!

Airblader commented:
Thanks for the confirmation. :)

psychon commented:
I wonder... should something like the following be done instead? (Which only occurred to me after having written that last commit)
```diff
diff --git a/Makefile.am b/Makefile.am
index 2117d2f..8ce6f26 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -34,8 +34,6 @@ libxcb_xrm_la_LDFLAGS = -version-info 0:0:0 -no-undefined -export-symbols-regex
 
 pkgconfig_DATA = xcb-xrm.pc
 
-if ENABLE_TESTS
-
 DIRECT_TESTS = tests/tests_parser tests/tests_match
 TESTS = $(DIRECT_TESTS) tests/tests_database_runner.sh
 check_PROGRAMS = $(DIRECT_TESTS) tests/tests_database
@@ -51,5 +49,3 @@ tests_tests_match_LDADD = libxcb-xrm.la $(XLIB_LIBS)
 tests_tests_database_SOURCES = tests/tests_utils.c tests/tests_database.c
 tests_tests_database_CPPFLAGS = -I$(srcdir)/include/ $(XCB_CFLAGS) $(XCB_AUX_CFLAGS)
 tests_tests_database_LDADD = libxcb-xrm.la $(XCB_LIBS) $(XCB_AUX_LIBS)
-
-endif
diff --git a/configure.ac b/configure.ac
index 0d9460b..0db3903 100644
--- a/configure.ac
+++ b/configure.ac
@@ -18,16 +18,17 @@ XCB_UTIL_COMMON([1.4], [1.6])
 PKG_CHECK_MODULES(XCB_AUX, xcb-aux)
 
 # The tests need Xlib, skip them if not found (unless --enable-tests is given)
-AC_ARG_ENABLE([tests],
-	      [AS_HELP_STRING([--disable-tests],
-			      [disable tests @<:@default=check@:>@])],
+AC_ARG_ENABLE([xlib-tests],
+	      [AS_HELP_STRING([--disable-xlib-tests],
+			      [disable xlib comparision tests tests @<:@default=check@:>@])],
 	      [],
-	      [enable_tests=check])
-AS_CASE(["$enable_tests"],
-	[yes], [PKG_CHECK_MODULES([XLIB], [x11 x11-xcb], [enable_tests=yes])],
-	[no], [enable_tests=no],
-	[PKG_CHECK_MODULES([XLIB], [x11 x11-xcb], [enable_tests=yes], [enable_tests=no])])
-AM_CONDITIONAL(ENABLE_TESTS, [test "x$enable_tests" = "xyes"])
+	      [enable_xlib_tests=check])
+AS_CASE(["$enable_xlib_tests"],
+	[yes], [PKG_CHECK_MODULES([XLIB], [x11 x11-xcb], [enable_xlib_tests=yes])],
+	[no], [enable_xlib_tests=no],
+	[PKG_CHECK_MODULES([XLIB], [x11 x11-xcb], [enable_xlib_tests=yes], [enable_xlib_tests=no])])
+AS_IF([test "x$enable_xlib_tests" = "xyes"],
+      [AC_DEFINE([XLIB_AVAILABLE])])
 
 AC_OUTPUT([Makefile
 	xcb-xrm.pc
diff --git a/tests/tests_match.c b/tests/tests_match.c
index 10cab1f..cfcad77 100644
--- a/tests/tests_match.c
+++ b/tests/tests_match.c
@@ -29,8 +29,11 @@
 #include <string.h>
 #include <limits.h>
 
+#ifdef XLIB_AVAILABLE
+#error
 #include <X11/Xlib-xcb.h>
 #include <X11/Xresource.h>
+#endif
 
 #include "tests_utils.h"
 
@@ -40,7 +43,9 @@ static int test_convert(void);
 static void setup(void);
 static void cleanup(void);
 
+#ifdef XLIB_AVAILABLE
 static char *check_get_resource_xlib(const char *str_database, const char *res_name, const char *res_class);
+#endif
 static int check_get_resource(const char *database, const char *res_name, const char *res_class, const char *value,
         bool expected_xlib_mismatch);
 static int check_convert_to_long(const char *value, const long expected, int expected_return_code);
@@ -233,6 +238,7 @@ static int test_convert(void) {
     return err;
 }
 
+#ifdef XLIB_AVAILABLE
 static char *check_get_resource_xlib(const char *str_database, const char *res_name, const char *res_class) {
     int res_code;
     char *res_type;
@@ -251,6 +257,7 @@ static char *check_get_resource_xlib(const char *str_database, const char *res_n
     XrmDestroyDatabase(database);
     return result;
 }
+#endif
 
 static int check_get_resource(const char *str_database, const char *res_name, const char *res_class, const char *value,
         bool expected_xlib_mismatch) {
@@ -258,7 +265,6 @@ static int check_get_resource(const char *str_database, const char *res_name, co
 
     bool err = false;
     char *xcb_value;
-    char *xlib_value;
 
     fprintf(stderr, "== Assert that getting resource <%s> / <%s> returns <%s>\n",
             res_name, res_class, value);
@@ -271,10 +277,12 @@ static int check_get_resource(const char *str_database, const char *res_name, co
         }
 
         if (!expected_xlib_mismatch) {
-            xlib_value = check_get_resource_xlib(str_database, res_name, res_class);
+#ifdef XLIB_AVAILABLE
+            char *xlib_value = check_get_resource_xlib(str_database, res_name, res_class);
             err |= check_strings(NULL, xlib_value, "Returned NULL, but Xlib returned <%s>\n", xlib_value);
             if (xlib_value != NULL)
                 free(xlib_value);
+#endif
         }
 
         goto done_get_resource;
@@ -284,12 +292,14 @@ static int check_get_resource(const char *str_database, const char *res_name, co
     free(xcb_value);
 
     if (!expected_xlib_mismatch) {
+#ifdef XLIB_AVAILABLE
         /* And for good measure, also compare it against Xlib. */
-        xlib_value = check_get_resource_xlib(str_database, res_name, res_class);
+        char *xlib_value = check_get_resource_xlib(str_database, res_name, res_class);
         err |= check_strings(value, xlib_value, "Xlib returns <%s>, but expected <%s>\n",
                 xlib_value, value);
         if (xlib_value != NULL)
             free(xlib_value);
+#endif
     }
 
 done_get_resource:
@@ -340,7 +350,11 @@ static int check_convert_to_bool(const char *value, const bool expected, const i
 }
 
 static void setup(void) {
+#ifdef XLIB_AVAILABLE
     XrmInitialize();
+#else
+    fputs("Xlib not available, skipping comparison\n", stderr);
+#endif
 }
 
 static void cleanup(void) {
```
What it does is to just skip the tiny "compare our result to Xlib"-part of `tests/match.c` instead of skipping all tests.

Airblader commented:
I'm generally somewhat against the idea of skipping tests. On the other hand I guess it does seem odd for an XCB library replacing Xlib functionality to pull Xlib, even if it's just for tests. Mhm.
